### PR TITLE
Modify sys.indexes view

### PR DESCRIPTION
**Pull request by ghost**
Signed-off-by: Avneet Toor <avneett@bitquilltech.com>

### Description

Because part of the work of `sys.indexes` was changing column data types, extensive work was done adding `sys.indexes` to the `babelfishpg_tsql--1.2.0--1.3.0.sql` upgrade script. However, PostgreSQL does not allow column data types to be changed in views once they exist, so any addition of updated views to the upgrade script where data types are changed returned an error. The steps I took to mitigate this was:
- in the upgrade script rename `sys.indexes` to `sys.indexes_deprecated`
- add the updated definition of `sys.indexes`
- add the definition of any views/functions that use `sys.indexes` as a dependency to rebuild them to use the updated `sys.indexes` definition. 
- if `sys.indexes_deprecated` has no other dependencies (by, say, a customer script) drop the view. Else, keep it for now. 

To aid in the above process I added 2 new helper functions:
- `sys.has_dependencies` : Returns True if an object is dependant upon by other objects. False otherwise.
- `sys.drop_view`: If a view is not depended upon by other objects, drop the view. Else, leave it. 
 
### Issues Resolved
BABELFISH-343
BABEL-2712

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--1.2.0--1.3.0.sql
Comment (kuntalghosh) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--1.2.0--1.3.0.sql]: The changes should be moved to 2.0.0-2.1.0 upgrade script.

Comment (ghost) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--1.2.0--1.3.0.sql]: Thank you, done. Now I need a maintainer to approve running workflows & 2 people to approve the PR.

File: contrib/babelfishpg_tsql/sql/sys_function_helpers.sql
Comment (nasbyj) [inline contrib/babelfishpg_tsql/sql/sys_function_helpers.sql]: s/by //

File: contrib/babelfishpg_tsql/sql/sys_function_helpers.sql
Comment (nasbyj) [inline contrib/babelfishpg_tsql/sql/sys_function_helpers.sql]: This function claims to be generic, but it is not. It will only work for views.

File: contrib/babelfishpg_tsql/sql/sys_function_helpers.sql
Comment (nasbyj) [inline contrib/babelfishpg_tsql/sql/sys_function_helpers.sql]: Function name sounds like it simply drops a view, which is not at all what it does.

**Review by nasbyj — CHANGES_REQUESTED** [review]
This is creating permanent catalog functions (aka *new product features*) to deal with a single one-off case. IMO that's not an appropriate way to handle this scenario. We should handle this via a simple plpgsql DO block that traps the exception we get from DROP VIEW if there are dependent objects. There's no need to dig into pg_depends.

Additionally, if we have to leave behind the deprecated view then we should disassociate it from the extension, as well as do a `RAISE WARNING` explaining to the user what's happening, and that it is their responsibility to drop the view when no longer needed. We should possibly move the view out of the sys schema as well.

**Comment by nasbyj** [comment]
One more comment: I absolutely do not think handling user objects that depend on catalog objects is a requirement. The Postgres community is very cautious about changing the definition of catalog objects, but when they feel it's warranted they simply change them and let the user deal with it.

File: contrib/babelfishpg_tsql/sql/sys_function_helpers.sql
Comment (ghost) [reply contrib/babelfishpg_tsql/sql/sys_function_helpers.sql]: As discussed offline, removed these permanent catalog functions in favour of creating a temporary procedure that serves the same purpose in the upgrade script which will then be dropped at the end.

File: contrib/babelfishpg_tsql/sql/sys_function_helpers.sql
Comment (ghost) [reply contrib/babelfishpg_tsql/sql/sys_function_helpers.sql]: As discussed offline, removed these permanent catalog functions in favour of creating a temporary procedure that serves the same purpose in the upgrade script which will then be dropped at the end. Hopefully the procedure name & description is more clear in the upgrade script.

File: contrib/babelfishpg_tsql/sql/sys_function_helpers.sql
Comment (ghost) [reply contrib/babelfishpg_tsql/sql/sys_function_helpers.sql]: As discussed offline, removed these permanent catalog functions in favour of creating a temporary procedure that serves the same purpose in the upgrade script which will then be dropped at the end. Hopefully the procedure name & description is more clear in the upgrade script.

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (nasbyj) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:826]: We shouldn't be doing that casting dance unless we have a really good reason to do so (and if so it needs to be documented)

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.0.0--2.1.0.sql
Comment (nasbyj) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.0.0--2.1.0.sql]: IMO this should be a WARNING

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (nasbyj) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:798]: We shouldn't be doing that regnamespace casting dance unless we have a really good reason to do so (and if so it needs to be documented)

**Review by nasbyj — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.0.0--2.1.0.sql
Comment (ghost) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.0.0--2.1.0.sql]: Done.

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (ghost) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:826]: I have created an issue to address this in the future here: https://github.com/babelfish-for-postgresql/babelfish_extensions/issues/190

This will take some time to change & properly test but I want to unblock other developers who require some of the additions in this PR ASAP. 

The view still works, and from the customer's perspective they won't see a difference if casting is in the implementation or not, but I do agree a change is in order and this casting stuff isn't great.

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (ghost) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:798]: I have created an issue to address this in the future here: https://github.com/babelfish-for-postgresql/babelfish_extensions/issues/190

This will take some time to change & properly test but I want to unblock other developers who require some of the additions in this PR ASAP. 

The view still works, and from the customer's perspective they won't see a difference if casting is in the implementation or not, but I do agree a change is in order and this casting stuff isn't great.

Comment (nasbyj) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:798]: Agreed, but please note the comment I just added to #190

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.0.0--2.1.0.sql
Comment (nasbyj) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.0.0--2.1.0.sql:9]: Just for reference sake...

I'm very tempted to say we should define this function in pg_temp, so that we ensure it can't be permanent. However, there's one significant gotcha to doing that: the extension will be marked as depending on this function, so if the function is sitting in pg_temp when the session disconnects, the forcible removal will actually drop the extension. Ultimately I wish there was a better way to safeguard/ensure that this function will actually be dropped at the end of the script...

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.0.0--2.1.0.sql
Comment (nasbyj) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.0.0--2.1.0.sql:13]: FWIW there's really no need for `query1` or `query2`; we could simply do `EXECUTE format($$....$$, schema_name, view_name)`.

**Review by nasbyj — DISMISSED** [review]

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.0.0--2.1.0.sql
Comment (ghost) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.0.0--2.1.0.sql:13]: Ack. Good to know.

File: contrib/babelfishpg_tsql/sql/sys_views.sql
Comment (ghost) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:798]: Ack

**Review by thephantomthief — APPROVED** [review]

**Review by KushaalShroff — APPROVED** [review]

**Pull request merged**
